components/bellsound/ConvertStep.tsx

```diff
@@ -0,0 +1,159 @@
+        return fileWithHeader
+    }
+
+    const getBestSampleRate = (lengthInSeconds: number, currentSampleRate: number): number => {
+        const maxFileSize = 400_000
+
+        const projectedCurrentSize = currentSampleRate * 2 * lengthInSeconds
+        if (projectedCurrentSize < maxFileSize) return currentSampleRate
+
+        const possibleSampleRates = [
+            44100,
+            22050,
+            16000,
+            11025,
+            8000,
+            6000
+        ]
+
+        for (const sampleRate of possibleSampleRates) {
+            const projectedSize = sampleRate * 2 * lengthInSeconds
+            if (projectedSize < maxFileSize) return sampleRate
+        }
+
+        return 6000
+    }
+
+    const startConversion = async () => {
+        if (converting) return
+        setConverting(true)
+
+        log("Choosing appropriate sample rate...")
+        const audioContext = new AudioContext()
+        const audioBuffer = await audioContext.decodeAudioData(await selectedFile.arrayBuffer())
+        const duration = audioBuffer.duration
+
+        const bestSampleRate = getBestSampleRate(duration, audioBuffer.sampleRate)
+
+        log(`Choosing sample rate of ${bestSampleRate} Hz based on file duration of ${duration} seconds`)
+
+        await loadFfmpeg()
+        const ffmpeg = globalFfmpeg
+
+        log("Loading file into memory...")
+        await ffmpeg.writeFile(selectedFile.name, await fetchFile(selectedFile))
+
+        log("Converting file...")
+        const ffmpegArgs = `-acodec pcm_s16le -ac 1 -ar ${bestSampleRate} -map_metadata -1 -fflags +bitexact`.split(" ")
+        const exitCode = await ffmpeg.exec(["-i", selectedFile.name, ...ffmpegArgs, "output.wav"])
+
+        if (exitCode !== 0) {
+            onError("There was an error converting your file. Please select a different one. Standard formats like MP3 or WAV work best.")
+            setConverting(false)
+            return
+        }
+
+        const convertedFile = await ffmpeg.readFile("output.wav", "binary") as Uint8Array
+
+        log("Applying VanMoof sound header...")
+        const fileWithHeader = applyHeader(convertedFile)
+
+        if (fileWithHeader.byteLength > 400_000) {
+            onError("The converted file is too large. Please select a shorter sound.")
+            setConverting(false)
+            return
+        }
+
+        onConversionCompleted(fileWithHeader)
+
+        log("Done!")
+        setConverting(false)
+    }
+
+    return (
+        <>
+            <p>Selected file: {selectedFile.name}</p>
+            {converting ?
+                <p>Converting...</p> :
+                <p>Now we need to convert your file to the VanMoof bell format. Press Convert to continue.</p>}
+            {showLog ? <textarea readOnly value={ffmpegLog}></textarea> : null}
+
+            <WalkthroughButton onClick={startConversion} disabled={converting} isPrimary>Convert</WalkthroughButton>
```

Comment:
Can you start this step automatically?
While we are converting the bell sound we can show why we need to convert the sound.

Reply:
I think we can start it automatically using `useEffect`, but I was worried that it might be called twice; was reading discussions online that it doesn't ever fire once reliably, even when configured to.

I guess we can stop it from starting the process again if it's already taking place, by checking the `converting` state.

The conversion process usually happens very quickly, so I'm not sure we should show any long text here if it starts automatically. We can put an explanation somewhere else though.